src/liballoc/rc.rs

```diff
@@ -570,15 +574,24 @@ impl<T: ?Sized> Rc<T> {
         ptr
     }
 
-    /// Constructs an `Rc` from a raw pointer.
+    /// Constructs an `Rc<T>` from a raw pointer.
     ///
-    /// The raw pointer must have been previously returned by a call to a
-    /// [`Rc::into_raw`][into_raw].
+    /// The raw pointer must have been previously returned by a call to
+    /// [`Rc<U>::into_raw`][into_raw] where `U` must have the same size
+    /// and alignment as `T`. This is trivially true if `U` is `T`.
+    /// Note that if `U` is not `T` but has the same size and alignment, this is
+    /// basically like transmuting references of different types. See
+    /// [`mem::transmute`][transmute] for more information on what
+    /// restrictions apply in this case.
     ///
-    /// This function is unsafe because improper use may lead to memory problems. For example, a
-    /// double-free may occur if the function is called twice on the same raw pointer.
+    /// The user of `from_raw` has to make sure a specific value of `T` is only
+    /// dropped once.
```

Comment:
What does this mean? Isn’t `Rc::drop` the one responsible for dropping `T` when the refcount reaches zero?

Reply:
The user is responsible for not calling `from_raw` (and allowing the returned value to drop) "too many times" is what this is getting at. I want to avoid saying something like "only once," since it can be incredibly useful to keep around a single pointer in FFI code and "clone out" Rc's from it, for example.

Reply:
Ok, I see what you’re getting at but I feel this one sentence does not express it well. Maybe we could talk about how each `Rc` or `Arc` “own” one strong reference, and `into_raw`/`from_raw` transfer ownership? I think that’s a better mental model, and double-drop is just one of the possible (bad) consequences.

(By the way I think there should be an alternative to `from_raw` that does *not* transfer ownership of the strong reference as a less fragile alternative to `ManuallyDrop::new(Rc::from_raw(ptr))`, but that’s out of scope for this PR. Maybe `&*mut T -> &Rc<T>`?)

Reply:
Agreed on both points. I'm not sure of the exact wording, but maybe @lukaslueg can come up with some draft wording?

Reply:
I'm a little hesitant on this point, for fear of convolution.

> Calling `into_raw()` / `from_raw()` does _not_ modify the internal reference counter, cloning and dropping the resulting `Rc<T>` does. One can think of these functions as temporarily lending exactly one "unit of ownership" to the raw pointer. The compiler can't check unique ownership through raw pointers, though, especially as all raw pointers are `Copy`. The user has to manually ensure `T` is always dropped exactly once, which is _not_ the case if (among other things) `T` is dropped through the raw pointer or if `from_raw()` is called multiple times on inadvertently made copies of the raw pointer.

Reply:
@Mark-Simulacrum any thoughts on the previous comment?

Reply:
I think it's certainly more detailed :)

However, I think it should clarify that even if you know T is safe to drop lots of times (e.g., `()` unit type), it is still not okay to let `Rc` drop more than once for the "last" time.